File: src/dotnet/Ansight.Tools.FileSystem/ReadFileTool.cs
namespace Ansight.Tools.FileSystem;

using System.Text.Json.Nodes;

public sealed class ReadFileTool : ITool
{
    private const int DefaultMaxBytes = 64 * 1024;
    private readonly FileSystemToolsOptions options;

    public ReadFileTool(FileSystemToolsOptions? options = null)
    {
        this.options = options ?? FileSystemToolsOptions.Default;
    }

    public string Category => "files";

    public ToolPolicy Policy => ToolPolicy.Read;

    public string Id => FileSystemToolIds.ReadFile;

    public string Name => "Read File";

    public string Description => "Reads a file from the app sandbox using a constrained path.";

    public string Keywords => "filesystem file read sandbox";

    public ToolSchema ArgumentsSchema => FileSystemToolSchemas.ReadFileArguments;

    public ToolSchema ResultSchema => FileSystemToolSchemas.ReadFileResult;

    public async Task<ToolResult> Execute(IReadOnlyDictionary<string, string> arguments)
    {
        ArgumentNullException.ThrowIfNull(arguments);

        try
        {
            var roots = FileSystemSandbox.GetRoots(options);
            var resolvedFile = FileSystemSandbox.ResolvePath(arguments, roots, requireExisting: true, expectDirectory: false);
            var maxBytes = FileSystemSandbox.GetInt(arguments, "maxBytes", defaultValue: DefaultMaxBytes, minimum: 1, maximum: 1024 * 1024);
            var encoding = FileSystemSandbox.GetString(arguments, "encoding");
            var fileInfo = new FileInfo(resolvedFile.FullPath);

            byte[] buffer;
            await using (var stream = new FileStream(resolvedFile.FullPath, FileMode.Open, FileAccess.Read, FileShare.Read))
            {
                var bytesToRead = (int)Math.Min(stream.Length, maxBytes);
                buffer = new byte[bytesToRead];
                var totalRead = 0;
                while (totalRead < bytesToRead)
                {
                    var bytesRead = await stream.ReadAsync(buffer.AsMemory(totalRead, bytesToRead - totalRead));
                    if (bytesRead <= 0)
                    {
                        break;
                    }

                    totalRead += bytesRead;
                }

                if (totalRead != buffer.Length)
                {
                    Array.Resize(ref buffer, totalRead);
                }
            }

            var truncated = fileInfo.Length > buffer.Length;
            var payload = FileSystemContentDescriptor.CreateResolvedFilePayload(resolvedFile, roots, fileInfo);
            payload["bytesRead"] = buffer.Length;
            payload["truncated"] = truncated;
            payload["capturedAtUtc"] = DateTime.UtcNow.ToString("O");

            var encodedContent = FileSystemContentDescriptor.EncodeContent(
                buffer,
                payload["mimeType"]!.GetValue<string>(),
                encoding);

            payload["contentType"] = encodedContent.ContentType;
            payload["encoding"] = encodedContent.Encoding;
            payload["text"] = encodedContent.Text;
            payload["base64"] = encodedContent.Base64;

            return ToolResult.Success(payload);
        }
        catch (Exception exception)
        {
            return ToolResult.Failure(exception.Message, errorCode: "filesystem_read_failed");
        }
    }
}

File: src/ios/Sources/AnsightToolsFileSystem/ReadFileTool.swift
import AnsightCore
import Foundation

public final class ReadFileTool: AnsightTool {
    private let options: AnsightFileSystemToolsOptions

    public init(options: AnsightFileSystemToolsOptions = .default) {
        self.options = options
    }

    public var descriptor: AnsightToolDescriptor {
        AnsightToolDescriptor(
            id: AnsightFileSystemToolIds.readFile,
            name: "Read File",
            description: "Reads a file from the app sandbox using a constrained path.",
            category: "files",
            policy: .read,
            keywords: "filesystem file read sandbox",
            argumentsSchema: AnsightFileSystemToolSchemas.readFileArguments,
            resultSchema: AnsightFileSystemToolSchemas.readFileResult
        )
    }

    public func execute(arguments: [String: String]) throws -> AnsightToolExecutionResult {
        do {
            let roots = try AnsightFileSystemSandbox.roots(options: options)
            let resolvedFile = try AnsightFileSystemSandbox.resolvePath(
                arguments: arguments,
                roots: roots,
                requireExisting: true,
                expectDirectory: false
            )
            let attributes = try AnsightFileSystemIO.attributes(path: resolvedFile.fullPath)
            let sizeBytes = Int64((attributes[.size] as? NSNumber)?.int64Value ?? 0)
            let maxBytes = try AnsightFileSystemSandbox.integer(
                arguments,
                key: "maxBytes",
                defaultValue: 64 * 1_024,
                minimum: 1,
                maximum: 1_024 * 1_024
            )
            let bytesToRead = Int(min(sizeBytes, Int64(maxBytes)))
            let data = try AnsightFileSystemIO.read(path: resolvedFile.fullPath, maxBytes: bytesToRead)
            let mimeType = AnsightFileSystemContentDescriptor.mimeType(path: resolvedFile.fullPath)
            let encoded = try AnsightFileSystemContentDescriptor.encodeContent(
                data: data,
                mimeType: mimeType,
                requestedEncoding: AnsightFileSystemSandbox.string(arguments, key: "encoding")
            )

            guard case .object(var payload) = AnsightFileSystemContentDescriptor.resolvedFilePayload(
                resolvedFile: resolvedFile,
                roots: roots,
                attributes: attributes
            ) else {
                return .failure("Failed to build file payload.", errorCode: "filesystem_read_failed")
            }

            payload["bytesRead"] = .integer(Int64(data.count))
            payload["truncated"] = .bool(sizeBytes > Int64(data.count))
            payload["capturedAtUtc"] = .string(AnsightClock.isoNow())
            payload["contentType"] = .string(encoded.contentType)
            payload["encoding"] = .string(encoded.encoding)
            payload["text"] = encoded.text.map(JSONValue.string) ?? .null
            payload["base64"] = encoded.base64.map(JSONValue.string) ?? .null
            return .success(.object(payload))
        } catch {
            return .failure(error.localizedDescription, errorCode: "filesystem_read_failed")
        }
    }
}
